File: 347 Top K Freq Elements/347.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        # If there's only one number in the list, return it
        if len(nums) == 1:
            return nums
        else:
            # Create a frequency map using Counter
            freq_map = Counter(nums)
            
            # Initialize a min heap
            min_heap = []
            
            # Iterate through each number and its frequency in the frequency map
            for num, freq in freq_map.items():
                # Push the (frequency, number) pair onto the min heap
                # We use frequency as the first element because heapq sorts based on the first element
                heapq.heappush(min_heap, (freq, num))

                # If the heap size exceeds k, remove the smallest element
                # This maintains the k most frequent elements in the heap
                if len(min_heap) > k:
                    heapq.heappop(min_heap)
            
            # Return a list of the k most frequent numbers
            # We only need the numbers, not their frequencies
            return [num for freq, num in min_heap]

File: 347 Top K Freq Elements/347.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        // If there's only one number in the vector, return it
        if (nums.size() == 1) {
            return nums;
        }
        else {
            // Create a frequency map
            unordered_map<int, int> freq_map;
            for (int num : nums) {
                freq_map[num]++;
            }

            // Define a min heap
            // We use greater<> to create a min heap instead of the default max heap
            priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> min_heap;

            // Iterate through each number and its frequency in the frequency map
            for (auto& [num, freq] : freq_map) {
                // Push the (frequency, number) pair onto the min heap
                min_heap.push({ freq, num });

                // If the heap size exceeds k, remove the smallest element
                // This maintains the k most frequent elements in the heap
                if (min_heap.size() > k) {
                    min_heap.pop();
                }
            }

            // Create a vector to store the k most frequent numbers
            vector<int> result;
            while (!min_heap.empty()) {
                result.push_back(min_heap.top().second);
                min_heap.pop();
            }

            return result;
        }
    }
};
